### ISSUE TITLE
Nesting `as_user` within `as_user` does not always set the audited user

### ISSUE THREAD

**Issue by KevinColemanInc** [collectiveidea/audited#449]
# Problem

`audited` does not support nesting `as_user` as it clears out `::Audited.store[:audited_user]` value at the end of `as_user`.

Guilty line:
https://github.com/collectiveidea/audited/blob/master/lib/audited/audit.rb#L137

# Use case

I want all audits done by a job to have `username` be`"job"` unless the job explicitly defines (via `as_user`) a certain user that triggered the job.   In sidekiq, I want to add middleware to set the `audited_user` to `"job"` and within the job, I sometimes know the user that triggered it, but using `as_user` breaks any audits that are created after the block.

```
user = User.first
Audited.audit_class.as_user('job') do
  user.update!(first_name: "job")

  Audited.audit_class.as_user(user) do
    user.update!(first_name: "me")
  end

  user.update!(first_name: "update by job")
end

puts user.audits.last(3).map { |u| [user_name: u.username, user_id: u.user_id] }

{:user_name=>"job", :user_id=>nil}
{:user_name=>nil, :user_id=>1}
{:user_name=>nil, :user_id=>nil}
```

The 2nd `as_user(user)` is nilling `::Audited.store[:audited_user]` resulting in the last update to not have a username or user associated with it.

# Proposed solution

My proposal is to have `as_user` restore the original value instead of nilling it.

```
def self.as_user(user, &block)
  last_audited_user = ::Audited.store[:audited_user]
  ::Audited.store[:audited_user] = user
  yield
ensure
  ::Audited.store[:audited_user] = last_audited_user
end
```

**Comment by KevinColemanInc** [comment]
I want `console`, `sidekiq`, and `rake` initiated audit usernames to default to their respectful values.  This is my "hack", since it seems `current_user` is only used by `sweeper` in the controller calls.

# Console & Rake
```ruby
# config/initializers/audited.rb
if defined?(Rails::Console)
  ::Audited.store[:current_user] = proc { :console }
elsif defined?(Rails::Server)
  # nothing; already handled by audited
else
  ::Audited.store[:current_user] = proc { :rake }
end
```

# Sidekiq 
```ruby
# app/middleware/sidekiq_audited.rb
class SidekiqAudited
  def call(_worker, _job, _queue)
    last_current_user = ::Audited.store[:current_user]
    ::Audited.store[:current_user] = proc { :sidekiq }
    yield
    ::Audited.store[:current_user] = last_current_user
  end
end

# config/initializers/sidekiq.rb
Sidekiq.configure_client do |config|
  config.server_middleware do |chain|
    chain.add SidekiqAudited
  end
end
```

**Comment by tbrisker** [comment]
@KevinColemanInc Your proposed solution makes sense to me, PRs welcome :wink:

**Comment by KevinColemanInc** [comment]
@tbrisker Done.  I just wanted to make sure it made sense to you so I don't waste any time. 👍

**Comment by tbrisker** [comment]
Solved by #450
